test/sanity/code-smell/update-bundled.py

```diff
@@ -0,0 +1,131 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+
+import glob
+import json
+import os
+import os.path
+import re
+import sys
+from distutils.version import LooseVersion
+
+import packaging.specifiers
+
+from ansible.module_utils.urls import open_url
+
+
+BUNDLED_RE = re.compile(b'\\b_BUNDLED_METADATA\\b')
+
+
+def get_bundled_libs(basedir):
+    bundled_libs = set()
+    for filename in glob.glob(os.path.join(basedir, 'lib/ansible/compat/*/__init__.py')):
+        if 'compat/tests' in filename:
+            # compat/tests doesn't bundle any code
+            continue
+        bundled_libs.add(filename)
+
+    bundled_libs.add(os.path.join(basedir, 'lib/ansible/module_utils/distro/__init__.py'))
+    bundled_libs.add(os.path.join(basedir, 'lib/ansible/module_utils/six/__init__.py'))
+    bundled_libs.add(os.path.join(basedir, 'lib/ansible/module_utils/compat/ipaddress.py'))
+    # backports.ssl_match_hostname should be moved to its own file
+    bundled_libs.add(os.path.join(basedir, 'lib/ansible/module_utils/urls.py'))
+
+    return bundled_libs
+
+
+def get_files_with_bundled_metadata(paths, basedir):
+    with_metadata = set()
+    for path in paths:
+        if path == 'test/sanity/code-smell/update-bundled.py':
+            continue
+
+        with open(path, 'rb') as f:
+            body = f.read()
+
+        if BUNDLED_RE.search(body):
+            with_metadata.add(os.path.join(basedir, path))
+
+    return with_metadata
+
+
+def get_bundled_metadata(filename):
+    with open(filename, 'r') as module:
+        for line in module:
+            if line.strip().startswith('_BUNDLED_METADATA'):
+                data = line[line.index('{'):].strip()
+                break
+        else:
+            raise ValueError('Unable to check bundled library for update.  Please add'
+                             ' _BUNDLED_METADATA dictionary to the library file with'
+                             ' information on pypi name and bundled version.')
+        metadata = json.loads(data)
+    return metadata
+
+
+def get_latest_applicable_version(pypi_data, constraints=None):
+    latest_version = "0"
+    if 'version_constraints' in metadata:
+        version_specification = packaging.specifiers.SpecifierSet(metadata['version_constraints'])
+        for version in pypi_data['releases']:
+            if version in version_specification:
+                if LooseVersion(version) > LooseVersion(latest_version):
+                    latest_version = version
+    else:
+        latest_version = pypi_data['info']['version']
+
+    return latest_version
+
+
+if __name__ == '__main__':
+    basedir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../..'))
```

Comment:
This is not needed. The code-smell tests are always executed by ansible-test from the root of the source tree. Using basedir results in absolute file paths, but test results should be relative paths. Instead of making them absolute and converting them back, simply use relative paths instead.

Reply:
I thought about that but thought that glob.glob() would need a full path.  (Maybe I was testing calling the script directly as well as through ansible-test).  Switching from glob.glob on the filesystem to operating on the paths passed in will remove the need for full paths as well.

Reply:
Done.